### PR TITLE
Handle Challenge or RandVal messages after disconnecting in Hello

### LINKED ISSUE #2150: KeysAuth decode_hex: Value must be an instance of str or bytes
### Symptomps:

in `golem.log`

```
2018-02-19 12:08:16 CRITICAL twisted                             b'Unhandled Error\nTraceback (most recent call last):\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\python\\log.py", line 103, in callWithLogger\n    return callWithContext({"system": lp}, func, *args, **kw)\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\python\\log.py", line 86, in callWithContext\n    return context.call({ILogContext: newCtx}, func, *args, **kw)\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\python\\context.py", line 122, in callWithContext\n    return self.currentContext().callWithContext(ctx, func, *args, **kw)\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\python\\context.py", line 85, in callWithContext\n    return func(*args,**kw)\n--- <exception caught here> ---\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\internet\\iocpreactor\\reactor.py", line 119, in _callEventCallback\n    evt.callback(rc, numBytes, evt)\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\internet\\iocpreactor\\abstract.py", line 92, in _cbRead\n    if self._handleRead(rc, data, evt):\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\internet\\iocpreactor\\abstract.py", line 118, in _handleRead\n    return self._dispatchData()\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\internet\\iocpreactor\\abstract.py", line 75, in _dispatchData\n    self.dataReceived(self._readBuffers[full_buffers][0:remainder])\n  File "C:\\Users\\ederenn\\Projects\\golem-env\\lib\\site-packages\\twisted\\internet\\iocpreactor\\tcp.py", line 75, in dataReceived\n    self.protocol.dataReceived(rbuffer)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\transport\\tcpnetwork.py", line 378, in dataReceived\n    self._interpret(data)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\transport\\tcpnetwork.py", line 405, in _interpret\n    self.session.interpret(m)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\p2p\\peersession.py", line 116, in interpret\n    BasicSafeSession.interpret(self, msg)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\transport\\session.py", line 75, in interpret\n    action(msg)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\p2p\\peersession.py", line 401, in _react_to_rand_val\n    self.__set_verified_conn()\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\p2p\\peersession.py", line 517, in __set_verified_conn\n    self.p2p_service.add_peer(self.key_id, self)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\network\\p2p\\p2pservice.py", line 295, in add_peer\n    self.keys_auth.get_difficulty(peer.key_id)\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\core\\keysauth.py", line 183, in get_difficulty\n    return int(math.floor(256 - math.log2(sha2(decode_hex(key_id)))))\n  File "C:\\Users\\ederenn\\Projects\\golem\\golem\\utils.py", line 58, in decode_hex\n    raise TypeError(\'Value must be an instance of str or bytes\')\nbuiltins.TypeError: Value must be an instance of str or bytes\n '
```

#### Encountered on
Windows 10
GOLEM Version: 0.11.0+dev399.g0bfff9e '
'Protocol Version: 23 '
'golem_messages Version: 1.9.0 '

### PR DESCRIPTION
**Pull request by etam**
Fixes #2150

### CODE REVIEW

**Comment by etam** [comment]
This is because `PeerSession._react_to_hello` sends `Hello` and `RandVal` or `ChallengeSolution` one after another. If a node decides to disconnect in `_react_to_hello` it'll still receive the second message.

File: golem/network/transport/session.py [hunk 1]
```diff
@@ -160,9 +161,10 @@ class BasicSafeSession(BasicSession):
     Cryptographic operation should be implemented in descendant class.
     """
 
+    key_id: Optional[str] = None
```
Comment (Krigpl) [inline golem/network/transport/session.py:161]: Pardon my ignorance but why is it better compared to `self.key_id =` in the constructor? Doesn't it make it possible to access it via `BasicSafeSession.key_id` which is useless (i.e. class vs object member)?

Comment (etam) [reply golem/network/transport/session.py:161]: https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations

Comment (Krigpl) [reply golem/network/transport/session.py:161]: wat

**Review by Krigpl — APPROVED** [review]

**Review by mfranciszkiewicz — CHANGES_REQUESTED** [review]
Resource handshakes time out and so does the DummyTask test

**Comment by etam** [comment]
@mfranciszkiewicz fixed

**Review by mfranciszkiewicz — APPROVED** [review]

**Review by jiivan — APPROVED** [review]

**Pull request merged**